### PR TITLE
Custom fake functions

### PR DESCRIPTION
**Pull request by gtaharaedmonds**
### Summary

Create new fake function framework (badly named as `fakegen` currently) since FFF and CMock don't meet our needs.

### Testing Done

Actually added unit tests (gasp) to make sure this actually works. Also added to CI.

### CODE REVIEW

**Comment by will-chaba** [comment]
had to run
`pipenv install pyclibrary`
and
`pipenv install jinja2`

Please document this for other users

**Comment by gtaharaedmonds** [comment]
> had to run `pipenv install pyclibrary` and `pipenv install jinja2`
> 
> Please document this for other users

Not necessary when using Docker.

**Comment by gtaharaedmonds** [comment]
Also plz suggest a better name `fakegen` sounds stupid

**Comment by will-chaba** [comment]
> Also plz suggest a better name `fakegen` sounds stupid

I'm telling you, FMock. It's not even that stupid, there's already CMock, GMock, lets do F(ormula)Mock

**Review by will-chaba — APPROVED** [review]
Looks good, will let you know if there are any additional features needed once I start porting the tests to this new framework, thanks for setting this up!

File: .vscode/launch.json [hunk 1]
```diff
@@ -115,6 +115,15 @@
...
+      "name": "Debug Tests: Shared",
+      "type": "cppdbg",
+      "request": "launch",
+      "program": "${workspaceRoot}/build_x86/firmware/shared/shared_test",
+      "cwd": "${workspaceRoot}",
+      "MIMode": "gdb",
+      "preLaunchTask": "Build: shared_test"
+    },
   ]
 }
```
Comment (will-chaba) [inline .vscode/launch.json:129]: In case you yerned for critique, add some new lines at the end of your files :neutral_face:

Comment (gtaharaedmonds) [reply .vscode/launch.json:129]: Lol github doesn't seem to like this, does it actually matter?

**Pull request merged**